### PR TITLE
caddyhttp: Make use of `http.ResponseController`

### PR DESCRIPTION
**Pull request by francislavoie**
One part of https://github.com/caddyserver/caddy/issues/5288, should fix https://github.com/caddyserver/caddy/issues/5431 (if built with Go 1.21+).

Adds a new `enable_full_duplex` server global option to opt-in (only effectual for Go 1.21+, with a `go:build` constraint). I think it needs to be opt-in because it could cause deadlocks if the client doesn't know how to handle a connection in that mode, but it seems necessary to turn on for certain kinds of apps. Maybe this should be an HTTP handler (maybe as a new option for https://caddyserver.com/docs/caddyfile/directives/request_body) which would allow using matchers to conditionally enable it instead of enabling it for all connections to the server :thinking: But I marked this as experimental so we can adjust later if needed.

The other main aspect of this is dropping `caddyhttp.HTTPInterfaces` because it's now made redundant by `http.ResponseController` which gives a way to use these optional interfaces without type assertions. Simplifies code in most cases.

Also syncs the reverseproxy implementation with stdlib's which now uses ResponseController as well; see golang/go@2449bbb for the original diff that I ported over.

And finally, minor improvement, `io.Copy()` already tries `io.ReaderFrom` so we don't need to do that ourselves in `responsewriter.go`.

### CODE REVIEW

**Comment by francislavoie** [comment]
Linter doesn't like this new pattern. Opened an issue: https://github.com/timakin/bodyclose/issues/52

Also the Go stdlib code we ported ignores the `err` returned by `m.flush()`, so I need to suppress it to keep the code the same :man_shrugging:

File: modules/caddyhttp/server.go [hunk 1]
```diff
@@ -82,6 +82,12 @@ type Server struct {
 	// HTTP request headers.
 	MaxHeaderBytes int `json:"max_header_bytes,omitempty"`
 
+	// Enable full-duplex communication for HTTP/1.1 requests.
+	// Only has an effect if Caddy was built with Go 1.21 or later.
```
Comment (mholt) [inline modules/caddyhttp/server.go:86]: Should we mention that this can brick/hang clients if they don't support this?

**Review by mholt — COMMENTED** [review]
Thanks, this definitely looks more elegant than what we were doing before.

I imagine there aren't external repos using our HTTPInterfaces type?

File: modules/caddyhttp/server.go (hunk 1 above)
Comment (francislavoie) [reply modules/caddyhttp/server.go:86]: Probably. But I'm not sure I understand well the circumstances where it can happen.

Comment (mholt) [reply modules/caddyhttp/server.go:86]: What if we say something like: "Test thoroughly with your client, as some clients may not support full-duplex HTTP/1.1 which can cause them to deadlock."

Comment (francislavoie) [reply modules/caddyhttp/server.go:86]: Done, also pulled in the text from Go stdlib which clarifies what it does.

**Review by mholt — APPROVED** [review]
LGTM now. Thank you for contributing this!

**Comment by francislavoie** [comment]
I'd just like @WeidiDeng to give his 👍 on this as well in case I missed something.

**Comment by WeidiDeng** [comment]
> I'd just like @WeidiDeng to give his 👍 on this as well in case I missed something.

There is something about `encode` that I don't like, can it be type asserted just like before to avoid wrapping a `caddyhttp.ResponseWriterWrapper`?

**Comment by francislavoie** [comment]
Hmm. I'm not sure what we would type assert at all. We already know it's a ResponseWriter. Do we even need wrapping now that response controller lets us call those optional interfaces in a unified way? Does our wrapper add any value, or is it necessary for it to work properly?

If you have a suggestion, feel free to commit it 🤔 I'm not sure I follow.

**Comment by francislavoie** [comment]
Aaah I see, so don't re-wrap if it's already wrapped. Makes sense.

**Comment by francislavoie** [comment]
Thanks, pushed that fix. Do you notice anything else?

**Comment by mholt** [comment]
I'll probably merge this in a few minutes to a few hours (before I have to take off for some family stuff today) so I can tag 2.7. Thank you for working on this!

**Comment by mholt** [comment]
And @WeidiDeng thank you for your invaluable review and contributions :pray:

**Pull request merged**